### PR TITLE
Fix playing videos in the radar widget

### LINKED ISSUE #20622: LUA: Media.PlayMovieInRadar crashes game [cnc: crash, ra: no function]
- Operating System:  Windows 10+11
- OpenRA Version: playtest-20230110 + release-20210321
- OpenRA Mod: cnc

lua.log:
```
A ScriptLobbyDropdown with ID `difficulty` was not found.
Fatal Lua Error: [string "-- Make copies of global functions to preve..."]:30: Uncaught CLR exception at Lua->CLR boundary: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.PlatformNotSupportedException: Operation is not supported on this platform.
   at OpenRA.Mods.Common.Scripting.MediaGlobal.AsyncLoader.BeginInvoke(Stream s, AsyncCallback callback, Object object)
   at OpenRA.Mods.Common.Scripting.MediaGlobal.PlayMovieInRadar(String movie, LuaFunction playComplete) in C:\Users\unXpected\Downloads\OpenRA-playtest-20230110\OpenRA-playtest-20230110\OpenRA.Mods.Common\Scripting\Global\MediaGlobal.cs:line 179
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Span`1& arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at OpenRA.Scripting.ScriptMemberWrapper.Invoke(LuaVararg args) in C:\Users\unXpected\Downloads\OpenRA-playtest-20230110\OpenRA-playtest-20230110\OpenRA.Game\Scripting\ScriptMemberWrapper.cs:line 76
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Span`1& arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Eluant.LuaRuntime.MethodWrapper.Invoke(Object[] parms)
   at Eluant.LuaRuntime.MakeManagedCall(IntPtr state, MethodWrapper wrapper)
   at OpenRA.Scripting.ScriptContext.FatalError(String message)
   at OpenRA.Mods.Common.Scripting.TriggerGlobal.<>c__DisplayClass2_0.<AfterDelay>b__0()
   at OpenRA.Effects.DelayedAction.<Tick>b__3_0(World w)
   at OpenRA.World.Tick()
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager)
   at OpenRA.Game.LogicTick()
   at OpenRA.Game.Loop()
   at OpenRA.Game.Run()
   at OpenRA.Game.InitializeAndRun(String[] args)
   at OpenRA.Launcher.Program.Main(String[] args)

```

I put the following in the map .lua:

Movie = function()
	Media.PlayMovieInRadar("gdi12.vqa")
end

and (in WorldLoaded)
Trigger.AfterDelay(DateTime.Seconds(3), function() Movie() end)

This immediately crashes the game every time. Tested with the playtest and release versions, cnc mod.
Media.PlayMovieFullscreen("gdi12.vqa") works without any issues.

In ra mod nothing happens, neither PlayMovieFullscreen nor PlayMovieInRadar do anything (also no errors and no entries in lua.log).

### PR DESCRIPTION
**Pull request by penev92**
Fixes #20622.

The original crash that was reported was because `Delegate.BeginInvoke()` is not supported on .NET Core and later, which I mention on the issue.

Diff probably looks scarier than it is.
~~This unified approach for playing videos from Lua removes the "async loading" from the radar videos (that wasn't used by the fullscreen videos!!), but I couldn't feel any severe lag so I hope that is fine.~~

### CODE REVIEW

**Review by PunkPun — COMMENTED** [review]
You have to have the videos installed for them to play correct? I don't have music / videos imported on my client and it didn't crash / do anything other than display messages in chat, so I guess it's working as intended.

But I'm concerned about the removal of async loading. IMO we should preserve and fix it, though for prep I think it's perfectly fine if we dropped it. 

Other than 👍

**Comment by penev92** [comment]
I don't think it's a big issue and I don't think we do it anywhere else (so no precedent), but I am open to suggestions about how to do it. It should be much lower-level than the Lua API though and ideally used for all assets everywhere.

**Review by abcdefg30 — DISMISSED** [review]
Not convinced that not drawing the overlay is a "fix", but this is good enough for prep. Thanks for tracking this down/completing the fix.

**Comment by penev92** [comment]
> Not convinced that not drawing the overlay is a "fix", but this is good enough for prep.

Not drawing the overlay is just ***"a"*** fix to a problem that came up while working on this. It has nothing to do with the original reported crash.

In any case I added one more point to the commit's description to make it clearer what it does and I added another commit adding "back" async video loading so we don't risk any delays when loading a video mid-game. We may actually want to consider switching fullscreen video playing to use the async method as well (but that is also used by victory/loss videos making it a bit more iffy).
I guess the new commit invalidates @abcdefg30's 👍  😕 

P.S.: Also added more test cases.

**Comment by penev92** [comment]
Rebased and removed `AsyncAction` as its only use-case was removed.

**Comment by PunkPun** [comment]
Fullscreen video executes OnComplete upon a failed video load attempt, while radar video doesn't

**Comment by penev92** [comment]
Fixed. Now they both call the callback function if the file is missing (we don't handle any other exceptions in either method, but that looks like it is by design).

File: OpenRA.Mods.Common/Scripting/Media.cs [hunk 1]
```diff
@@ -60,27 +59,11 @@ public static void PlayFMVFullscreen(World w, string movie, Action onComplete)
...
-		public static void StopFMVInRadar()
-		{
-			var player = Ui.Root.Get<VideoPlayerWidget>("PLAYER");
-			player.Stop();
-		}
-
-		public static IVideo LoadVideo(Stream s)
-		{
-			return VideoLoader.GetVideo(s, true, Game.ModData.VideoLoaders);
+			player.DrawOverlay = false;
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Scripting/Media.cs]: I still don't understand this. It's never set to true, why is it here? Why is it needed? An explanation should be added to the variable

Comment (penev92) [reply OpenRA.Mods.Common/Scripting/Media.cs]: The default value is `true`: https://github.com/OpenRA/OpenRA/blob/5bf7fe852c8f122e3a3e0a1eb72d66d7429675c3/OpenRA.Mods.Common/Widgets/VideoPlayerWidget.cs#L24

I guess there was another crash hiding behind the `.BeginInvoke()` one, so I added that line on a hunch (which paid off). Investigating it further now, I guess the issue is related to this:
video_played_widget_draw_overlay
I have a better fix that I will push tonight after work.

**Comment by penev92** [comment]
Updated.

**Comment by penev92** [comment]
Added back video cancelling by another video in the last commit.

**Comment by penev92** [comment]
Also added a race condition fix for when loading 2 videos for the radar widget at the same time
and snuck in the addition of a full-screen in-gave video player background for D2k as extra polish.

**Review by PunkPun — DISMISSED** [review]
LGTM

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/f9c71461ec340cc4af6e10cf21fff37ae32be387...375e776baf215f68e52a1c9bce97c3f8e8ab02d3)

https://github.com/OpenRA/OpenRA/commit/44f35729e117c0476071b50002648c5b9cc99aff
https://github.com/OpenRA/OpenRA/commit/d7b2727d16a147828f544b7ce3799377c050ab0f
https://github.com/OpenRA/OpenRA/commit/d99991d317ec2c95d3802555e095fea103c3480f
https://github.com/OpenRA/OpenRA/commit/1f63294503bae868650386c679a424cc66ddb6df
https://github.com/OpenRA/OpenRA/commit/7bc2768dc4cea18f7ca8b1013232bde7d80af6ae
https://github.com/OpenRA/OpenRA/commit/7867eb99c6e48498ac61a3f7cc3737c1fded7728

**Pull request merged**